sql-plugin/src/main/scala/com/nvidia/spark/rapids/optimizer/FactDimensionJoinReorder.scala

```diff
@@ -0,0 +1,439 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.optimizer
+
+import java.util.concurrent.TimeUnit
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+import com.nvidia.spark.rapids.RapidsConf
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, AttributeReference, BinaryExpression, Expression, ExpressionSet, IsNotNull, PredicateHelper}
+import org.apache.spark.sql.catalyst.plans.Inner
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+import org.apache.spark.sql.types.{DataTypes, StructType}
+
+/**
+ * Join reordering rule based on the paper "Improving Join Reordering for Large Scale Distributed
+ * Computing", available at https://ieeexplore.ieee.org/document/9378281.
+ *
+ * This optimizer rule is not specific to GPU and could potentially be contributed back to Apache
+ * Spark.
+ */
+object FactDimensionJoinReorder 
+    extends Rule[LogicalPlan] 
+    with PredicateHelper 
+    with Logging {
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    val conf = new RapidsConf(plan.conf)
+    if (!(conf.isSqlEnabled && conf.joinReorderingEnabled)) {
+      return plan
+    }
+    val t0 = System.nanoTime()
+    val reorderedJoin = plan.transformUp {
+      case j@Join(_, _, Inner, Some(_), JoinHint.NONE) if isSupportedJoin(j) =>
+        reorder(j, conf)
+      case p @ Project(projectList, Join(_, _, Inner, Some(_), JoinHint.NONE))
```

Comment:
nit remove spaces p@Project or put spaces in j @ Join to keep consistent

Reply:
I thought TransformUp looks at children first so would we have already reordered the Join which is a child of this Project?  I might be not remembering the exact flow there though.  we don't need to check anything about the Join that is children like isSupportedJoin?

Reply:
> I thought TransformUp looks at children first

Yes, that is correct. The intent is to apply the optimization to children first because we can sometimes optimize part of a join but not the whole join.

> so would we have already reordered the Join which is a child of this Project?

That's a good question and I will experiment with removing this and see what happens. This code is based on the code in Spark's `CostBasedJoinReorder` but that was using `transformDown` so maybe this is redundant here.